extension/src/minn_maxn.rs

```diff
@@ -0,0 +1,1177 @@
+use pgx::{iter::SetOfIterator, *};
+
+use serde::{Deserialize, Serialize};
+
+use crate::{
+    aggregate_utils::in_aggregate_context,
+    flatten,
+    palloc::{Inner, Internal, InternalAsValue, ToInternal},
+    pg_type,
+    raw::bytea,
+    ron_inout_funcs,
+};
+
+use ordered_float::NotNan;
+
+use std::cmp::Reverse;
+use std::collections::BinaryHeap;
+use std::marker::PhantomData;
+
+#[derive(Clone, Serialize, Deserialize)]
+pub struct NMostTransState<T: Ord> {
+    capacity: usize,
+    heap: BinaryHeap<T>,
+    phantom: PhantomData<T>,
+}
+
+impl<T: Ord> NMostTransState<T> {
+    fn new(capacity: usize, first_val: T) -> NMostTransState<T> {
+        let mut new_heap = NMostTransState {
+            capacity,
+            heap: BinaryHeap::with_capacity(capacity),
+            phantom: PhantomData,
+        };
+
+        new_heap.new_entry(first_val);
+
+        new_heap
+    }
+
+    fn new_entry(&mut self, new_val: T) {
+        // If at capacity see if we need to replace something
+        if self.heap.len() == self.capacity {
+            if !self.belongs_in_heap(&new_val) {
+                return;
+            }
+
+            self.heap.pop();
+        }
+
+        self.heap.push(new_val)
+    }
+
+    fn belongs_in_heap(&self, val: &T) -> bool {
+        // Note that this will actually be '>' if T is a Reverse<...> type
+        val < self.heap.peek().unwrap()
+    }
+}
+
+impl<T: Ord + Copy> From<&[T]> for NMostTransState<T> {
+    fn from(vals: &[T]) -> Self {
+        let mut state = Self::new(vals.len(), vals[0]);
+        for val in vals[1..].iter() {
+            state.new_entry(*val);
+        }
+        state
+    }
+}
+
+fn min_max_trans_function<T>(
+    state: Option<Inner<NMostTransState<T>>>,
+    val: T,
+    capacity: usize,
+    fcinfo: pg_sys::FunctionCallInfo,
+) -> Option<Inner<NMostTransState<T>>>
+where
+    T: Ord,
```

Comment:
nit: this looks clearer without a `where` clause
```suggestion
fn min_max_trans_function<T: Ord>(
    state: Option<Inner<NMostTransState<T>>>,
    val: T,
    capacity: usize,
    fcinfo: pg_sys::FunctionCallInfo,
) -> Option<Inner<NMostTransState<T>>>
```

Reply:
Fixed